ARCHITECTURE.md

```diff
@@ -0,0 +1,357 @@
+primary goal is to enable M:N concurrency in Rust programs; multiple coroutines
+can be multiplexed onto a single thread, allowing a server to handle hundreds
+of thousands (if not millions) of clients at a time. However, [`smol`] can just
+as easily multiplex tasks onto multiple threads to enable blazingy fast
+concurrency. [`smol`] is intended to work on any scale; [`smol`] should work for
+programs with two coroutines as well as two million.
+
+On an architectural level, [`smol`] prioritizes maintainable code and clarity.
+[`smol`] aims to provide the performance of a modern `async` runtime while still
+remaining hackable. This philosophy informs much of the decisions in [`smol`]'s
+codebase and differentiate it from other contemporary runtimes.
+
+On a technical level, [`smol`] is a [work-stealing] executor built around a
+[one-shot] asynchronous event loop. It also contains a thread pool for
+filesystem operations and a reactor for waiting for child processes to finish.
+[`smol`] itself is a meta-crate that combines the features of numerous subcrates
+into a single `async` runtime-based package.
+
+smol-rs consists of the following crates:
+
+- [`async-io`] provides a one-shot reactor for polling asynchronous I/O. It is
+  used for registering sockets into `epoll` or another system, then polling them
+  simultaneously.
+- [`blocking`] provides a managed thread-pool for polling blocking operations as
+  asynchronous tasks. It is used by many parts of [`smol`] for turning
+  operations that would normally be non-concurrent into concurrent operations,
+  and vice versa.
+- [`async-executor`] provides a work-stealing executor that is used as the
+  scheduler for an `async` program. While the executor isn't as optimized as
+  other contemporary executors, it provides a performant executor implemented in
+  mostly safe code in under 1.5 KLOC.
+- [`futures-lite`] provides a handful of low-level primitives for combining and
+  dealing with `async` coroutines.
+- [`async-channel`] and [`async-lock`] provide synchronization primitives that
+  work to connect asynchronous tasks.
+- [`async-net`] provides a set of higher-level APIs over networking primitives.
+  It combines [`async-io`] and [`blocking`] to create a full-featured, fully
+  asynchronous networking API.
+- [`async-fs`] provides an asynchronous API for manipulating the filesystem. The
+  API itself is built on top of [`blocking`].
+
+These subcrates in and of themselves depend on subcrates for further
+functionality. These are explained in a more bottom-up fashion below.
+
+## Lower Level Crates
+
+These crates provide safer, lower-level functionality used in the higher-level
+crates. These could be used in higher-level crates but are intended primarily
+for use in [`smol`]'s underlying plumbing.
+
+### [`parking`]
+
+[`parking`] is used to block threads until an arbitrary signal is received, or
+"parks" them. The [`std::thread::park`] API suffers from involving global state;
```

Comment:
", or parks them" is worded a bit confusingly (as "parks" is/should be associated to "block threads"), but this isn't obvious here.